### PR TITLE
[Bugfix] Fix PD disagg + MTP correctness for Qwen3.5(GDN)

### PR DESCRIPTION
**Pull request by andakai**
## Problem

`Qwen3.5-0.8B` with `MTP=1` can lose GSM8K accuracy under PD disaggregation when concurrency is greater than 1 and default cudagraph is enabled. The issue was reproduced with both Mooncake and NIXL.

Trigger pattern:

- model: `Qwen3.5-0.8B`
- hybrid/GDN state model
- PD disaggregation
- MTP/spec decode enabled, `num_speculative_tokens=1`
- TP=1
- concurrent requests, so decode can batch normal running requests with a newly handed-off request
- default cudagraph enabled

| Connector | Config | Accuracy |
|---|---|---:|
| Mooncake | before fix, 4P4D | `0.153904` |
| NIXL | before fix, 4P4D | `0.185747` |

## Root Cause

**On the decode worker, a newly handed-off GDN request will not have completed the last prompt token**, because the decode worker still needs to forward the prompt tail token `x_N` to advance GDN state to `h(N)`.

So in D instance:
```text
num_computed_tokens = num_prompt_tokens - 1
state = h(N-1)
```

In the previous version, D instance will go into this branch, because its `num_new_tokens=1`. **However, this is actually the prompt tail token, so this prefill token will go into the decode.**
```
if (
    (self.num_spec_tokens > 0 and self.dynamic_sd_lookup is None)
    and num_new_tokens == 1
    and num_computed_tokens >= request.num_prompt_tokens
    and (scheduled_running_reqs and not prefill_scheduled)
):
    num_new_tokens = 1 + self.num_spec_tokens
```

Old scheduler logic could still pad this boundary step as MTP decode:

```text
input = [xN, -1]
query_len = 1 + mtp_tokens
```

The input assembly kernel will execute the decode rewrite in `_combine_sampled_and_draft_tokens_kernel`

```python
if NUM_NEW_SAMPLED_TOKENS > 0:
  # Write the last sampled token ID to input_ids.
  last_token_id = tl.load(last_sampled_tokens_ptr + req_state_idx)
  tl.store(input_ids_ptr + query_end - num_logits, last_token_id)
```

For normal decode this is correct,  this is used to write the previously sampled/accepted token into the decode input slot for the next forward pass, **and the model should consume the last generated token, not another prompt token**. In this PD+GDN situation, it is the previous computed token, `x(N-1)`. So the prompt tail `xN` is overwritten before GDN forward, and the recurrent state advances from the wrong token.

This explains the accuracy collapse: once GDN state is corrupted at the PD handoff boundary, later generation follows the wrong state trajectory.

## Fix

Only schedule MTP padding for a request after its prompt has completed locally. In this way, the last token in D instance will be treated as a prefill token and does not overwrite `xN`.

```diff
 if (
     (self.num_spec_tokens > 0 and self.dynamic_sd_lookup is None)
     and num_new_tokens == 1
+    and num_computed_tokens >= request.num_prompt_tokens
     and (scheduled_running_reqs and not prefill_scheduled)
 ):
     num_new_tokens = 1 + self.num_spec_tokens
```

## Validation

All accuracy numbers below use `Qwen3.5-0.8B`, GSM8K full set, 5-shot, `max_tokens=256`, `MTP=1`, `TP=1`. Use default config.

| Connector | Config | Accuracy |
|---|---|---:|
| Mooncake | before fix, 4P4D | `0.153904` |
| Mooncake | after fix, 4P4D | `0.321456` |
| NIXL | before fix, 4P4D | `0.185747` |
| NIXL | after fix, 4P4D | `0.323730` |

This is close to the standalone accuracy.
---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

**Comment by andakai** [comment]
Hi, @qianlihuang . I encountered an accuracy issue when using PD+MTP+GDN. I think it is related to your pr https://github.com/vllm-project/vllm/pull/45237. In your previous test, if `num_computed_tokens == request.num_prompt_tokens-1` can also have spec padding and do uniform decoding. This will cause the issue in this pr, so I added this logic. 

Do you think this will influence your previous pr? In my opinion, this pr only influences the cases that have `num_computed_tokens == request.num_prompt_tokens-1`. Do you think this can be acceptable?

**Comment by andakai** [comment]
After discussing with @qianlihuang , I realized that the previous fix will influence the previous pr, so I find another way to fix this issue. Here are two fixes.

The change keeps the scheduler-side MTP padding optimization for PD/prefix-cache boundary steps, so a boundary-shaped batch like [xN, -1] can still use the uniform 1 + spec_tokens spec-decode/cudagraph path. 

1. The input assembly kernel is tightened so `last_sampled_token` only rewrites real generated-token slots, preventing the prompt-tail token xN from being overwritten at the handoff boundary.

2. Previously, a PD handoff batch like [xN, -1] had the shape of spec decode but was not treated as GDN spec metadata, so GDN/SSM state could be updated with the wrong token/state semantics. The fix routes these spec-shaped boundary batches through the existing GDN spec metadata path, where num_accepted_tokens controls how many tokens actually update conv/SSM state. This reuses the existing GDN spec path and prevents -1 padding from corrupting state.

**Comment by njhill** [comment]
Thanks @andakai, this looks good to me. However I think we need equivalent fix for the MRV2 SSM spec decode condition. I have pushed a commit with that along with removing the now-redundant `last_sampled_tokens` seeding. Please take a look and lmk what you think. cc @qianlihuang

**Comment by andakai** [comment]
Thanks for the follow-up fix! This also looks good to me!

**Comment by andakai** [comment]
I checked the three failing CI logs and they appear unrelated to this PR.

**Review by njhill — APPROVED** [review]
Thanks @andakai
